### PR TITLE
Track pending dials

### LINKED ISSUE #4790: Network discovery: lodestar failed to dial a lot of peers at the same time
**Describe the bug**

Found this on a mainnet nodes running v1.2.1, lodestar dials same peer at the same time

```
beacon-2022-11-21.log:366296:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were rejecte
beacon-2022-11-21.log:366298:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were reject
beacon-2022-11-21.log:366300:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were rejec
beacon-2022-11-21.log:366302:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were reje
beacon-2022-11-21.log:366304:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were rej
beacon-2022-11-21.log:366306:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were re
beacon-2022-11-21.log:366308:Nov-21 10:24:11.752[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were r
beacon-2022-11-21.log:366310:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises were
beacon-2022-11-21.log:366312:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises wer
beacon-2022-11-21.log:366314:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises we
beacon-2022-11-21.log:366316:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises w
beacon-2022-11-21.log:366318:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promises
beacon-2022-11-21.log:366320:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promise
beacon-2022-11-21.log:366322:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promis
beacon-2022-11-21.log:366324:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All promi
beacon-2022-11-21.log:366326:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All prom
beacon-2022-11-21.log:366328:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All pro
beacon-2022-11-21.log:366330:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All pr
beacon-2022-11-21.log:366332:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All p
beacon-2022-11-21.log:366334:Nov-21 10:24:11.753[network]         debug: Error dialing discovered peer peer=16...WSFRXL All
```

it also dials different peers
```
beacon-2022-11-21.log:365402:Nov-21 10:23:50.407[network]         debug: Error dialing discovered peer peer=16...LaRaTu All promises were rejecte
beacon-2022-11-21.log:365420:Nov-21 10:23:50.823[network]         debug: Error dialing discovered peer peer=16...XJKdXg All promises were rejecte
beacon-2022-11-21.log:365516:Nov-21 10:23:51.128[network]         debug: Error dialing discovered peer peer=16...SySh89 All promises were rejecte
beacon-2022-11-21.log:365550:Nov-21 10:23:51.427[network]         debug: Error dialing discovered peer peer=16...XJKdXg All promises were rejecte
beacon-2022-11-21.log:365552:Nov-21 10:23:51.427[network]         debug: Error dialing discovered peer peer=16...XJKdXg All promises were reject
beacon-2022-11-21.log:365689:Nov-21 10:23:55.143[network]         debug: Error dialing discovered peer peer=16...SySh89 All promises were rejecte
```

this was found when investigating #4661 when we receive a lot of goodbye requests

Screen Shot 2022-11-21 at 19 25 57

this error happens a lot

Screen Shot 2022-11-21 at 19 26 33

**Expected behavior**

Lodestar should not keeps dialing same peer at the same time

### PR DESCRIPTION
**Pull request by wemeetagain**
- Ensure that a peer isn't already being dialed before attempting to dial again
- Fix the formatting bug https://gist.github.com/epheph/5ffc9e04b927ea5a892b2f6e001ab383

- Resolves https://github.com/ChainSafe/lodestar/issues/4790

### CODE REVIEW

File: packages/beacon-node/src/network/peers/discover.ts [hunk 1]
```diff
@@ -82,7 +83,7 @@ export class PeerDiscovery {
   private metrics: IMetrics | null;
   private logger: ILogger;
   private config: IBeaconConfig;
-  private cachedENRs = new Set<CachedENR>();
+  private cachedENRs = new Map<PeerIdStr, CachedENR>();
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/peers/discover.ts:86]: First change is to cache ENRs indexed by peer id

File: packages/beacon-node/src/network/peers/discover.ts [hunk 2]
```diff
@@ -279,6 +286,11 @@ export class PeerDiscovery {
         return DiscoveredPeerStatus.already_connected;
       }
 
+      // Ignore dialing peers
+      if (getDefaultDialer(this.libp2p).pendingDials.has(peerId.toString())) {
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/peers/discover.ts:290]: Second change is to check `libp2p.dialer.pendingDials` before dialing

File: packages/beacon-node/src/network/peers/discover.ts [hunk 3]
```diff
@@ -493,7 +510,8 @@ export class PeerDiscovery {
  */
 function formatLibp2pDialError(e: Error): void {
   const errorMessage = e.message.trim();
-  e.message = errorMessage.slice(0, errorMessage.indexOf("\n"));
+  const newlineIndex = errorMessage.indexOf("\n");
+  e.message = newlineIndex !== -1 ? errorMessage.slice(0, newlineIndex) : errorMessage;
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/peers/discover.ts:514]: Third change is to fix this error formatter

File: packages/beacon-node/src/network/util.ts [hunk 4]
```diff
@@ -81,3 +84,7 @@ export function getConnection(connectionManager: ConnectionManager, peerIdStr: s
 export function isPublishToZeroPeersError(e: Error): boolean {
   return e.message.includes("PublishError.InsufficientPeers");
 }
+
+export function getDefaultDialer(libp2p: Libp2p): DefaultDialer {
+  return ((libp2p as unknown) as {components: Components}).components.dialer as DefaultDialer;
```
Comment (twoeths) [inline packages/beacon-node/src/network/util.ts:89]: looks like a work-around if libp2p doesn't public the DefaultDialer
@wemeetagain do you know if latest libp2p version has an official api to get this?

**Review by twoeths — APPROVED** [review]
looks good to me, dropped a minor comment

File: packages/beacon-node/src/network/util.ts (hunk 4 above)
Comment (wemeetagain) [reply packages/beacon-node/src/network/util.ts:89]: There's no 'official' way to get this. They only make the `Dialer` interface public. Might be nice to expose the type in a nicer way :shrug:

**Pull request merged**